Feature autojournal: allow user to automatically generate journal entries from commands/moves

martinellison opened the pull request:
This is a work-in-progress draft for review; as previously mentioned there are several questions:

1. This proposal is larger than previous changes and changes the architecture more. So we need to discuss whether this is the correct approach. I would argue that, although this change introduces complexity, it is the simplest way to retain the existing functionality and to add the proposed functionality.
2. The output that this generates to the journal file is not correct yet: some information that is shown on the console does not get written to the journal, and some output written to the journal is not required (e.g. the 'journal' command gets logged, then it writes the same data to the journal again). 

I am putting this change up for comment and review at this stage; I will need to do further work before submitting the change for approval.

thexhr commented:
Hi,

1. I find the difference between the "journal" and the "startjournal" command quite confusing. What do you think about changing the latter to "startautojournal"? Same for the "stop" command?
2. The PR contains some changes that are not directly related to the new code but more general changes. Just submit these as separate PRs, I can review and merge them easy and it keeps this PR clean. I left you some review comments in the code.
3. I am not quite happy with the regex code. This is a lot of code just to remove the control characters. My proposal would be to get a version ready which only logs plain text. Once that code is working and community tested, the color can be tackled. Can you modify the PR accordingly?

martinellison commented on isscrolls.c:
```diff
@@ -321,15 +328,71 @@ get_cursed(void)
+clear_message_buffer(void) {
+	message_buffer_pos = &message_buffer[0];
+	buffer_chars_left = BUFFER_LENGTH;
+	message_buffer[0] = '\0';
```
coding

martinellison commented on isscrolls.c:
```diff
@@ -321,15 +328,71 @@ get_cursed(void)
+	int chars_written;
+	log_debug("writing to buffer, %d chars left, format is '%s'\n", buffer_chars_left, format);
+	va_start(args, format);
+    chars_written = vsprintf(message_buffer_pos, format, args);
```
coding; sorry, missed that one; I'm trying to get the Visual Studio Code settings correct for tabs

martinellison commented on notes.c:
```diff
@@ -115,6 +115,7 @@ select_note(char *cmd)
 	}
 
 	return nid;
+
```
removing

martinellison commented on notes.c:
```diff
@@ -380,6 +381,7 @@ load_note(int nid, struct note *n)
 			curchar->id == json_object_get_int(id)) {
 			log_debug("Loading note for id: %d\n", json_object_get_int(lid));
 
+
```
removing

martinellison commented on character.c:
```diff
@@ -1346,7 +1350,7 @@ init_character_struct(void)
 		log_errx(1, "calloc");
 
-	if ((c->vow= calloc(1, sizeof(struct vow))) == NULL)
+	if ((c->vow = calloc(1, sizeof(struct vow))) == NULL)
```
removing

martinellison commented:
> 1. I find the difference between the "journal" and the "startjournal" command quite confusing. What do you think about changing the latter to "startautojournal"? Same for the "stop" command?

changing

martinellison commented:
I have made changes for the specific comments, and also edited the `man` page to describe autojournalling.

> 3\. I am not quite happy with the regex code. This is a lot of code just to remove the control characters. My proposal would be to get a version ready which only logs plain text. Once that code is working and community tested, the color can be tackled. Can you modify the PR accordingly?

Ok, at the moment (as of say release `2024a`), `isscrolls` writes messages (let's call these 'command output') to standard output which, depending on the command line options, may contain colored text. As of HEAD, `isscrolls` also writes entries to the journal file (without color).

What we want to do in the new version is to write some of the command output to the journal as well as to standard output. But the text sent to the journal should not be in color (that is, it should not contain any set/clear color binary sequences). How do we do this?

The solution that I have coded so far creates a single version of the command output, including color binary sequences, and then removes the color sequences from the version sent to the journal. This solution does create the correct results, but is difficult to understand.

One alternative is to remove the color altogether. This is a much simpler solution, but it removes the color functionality. Is this what you are proposing?

thexhr commented:
> edit 2: also, I could merge pm() and add_to_buffer() into a single function, and pass a parameter to set the color (for the console) and also control whether the message goes to the console, the journal, or both. That would be simpler.

That was exactly my thought. There is a single `vprintf()` in the `pm()` function that writes the actual non-colored text to stdout. Here you jump in and construct the message buffer

thexhr commented on character.c:
```diff
@@ -1421,11 +1441,11 @@ cmd_journal(char *what)
 			goto again;
 		}
-		snprintf(entry, MAX_ENTRY_LEN, "%s", prompted);
+		snprintf(entry, MAX_ENTRY_LEN, "%s\n", prompted);
```
Is that extra \n here intentional? In my tests the journal here looks as follows:

```
[2025-03-28 13:22:36] undertakeajourney <2> + 1 = 3 vs <6><4> miss

[2025-03-28 13:22:38] undertakeajourney <5> + 1 = 6 vs <4><7> weak hit

[2025-03-28 13:22:40] undertakeajourney <6> + 1 = 7 vs <2><8> weak hit

[2025-03-28 13:22:43] q
```

I would prefer one entry per line without any additional extra blank lines

martinellison replied:
this should be fixed now

thexhr commented on isscrolls.c:
```diff
@@ -321,25 +331,45 @@ get_cursed(void)
+}
+
+int
+journaling() {
```
Please change to `journaling(void)` to make recent versions of clang happy.

martinellison replied:
coding

thexhr commented on isscrolls.c:
```diff
@@ -321,25 +331,45 @@ get_cursed(void)
 
 void
-write_journal_entry(char const * const what)
+start_journal_entry()
```
Same as above: `start_journal_entry(void)`

martinellison replied:
coding

thexhr commented on rolls.c:
```diff
@@ -19,6 +19,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <stdarg.h>
```
This include can be removed

martinellison replied:
coding

thexhr commented on rolls.c:
```diff
@@ -671,6 +672,7 @@ action_roll(int args[2])
 	long c1, c2, a1, b, cd;
 	int ret = 0, match = 0;
 
+
```
Superfluous blank line

martinellison replied:
coding

thexhr commented on isscrolls.c:
```diff
@@ -367,3 +397,4 @@ close_journal_file(void)
 	}
 }
 
+
```
Superfluous blank line

martinellison replied:
coding

thexhr commented on isscrolls.1:
```diff
@@ -152,6 +152,10 @@ In case you have an active vow, its title and description will be shown.
 It is possible to create notes for future reference.
 .Pp
 It is also able to write entries to a journal file.
+.Pp
```
Don't forget to bump the date at the beginning.

martinellison replied:
coding

thexhr commented:
Looks quite good now. I left some smaller comments. Once they are resolved I'll merge the PR.

martinellison commented:
If you do not have any major objections, I would want to make some more changes before I submit the pull request, to:

- fix the newlines in the journal as discussed above (https://github.com/thexhr/isscrolls/pull/18#discussion_r2018712708)
- suppress some commands from being journaled
- convert some more command responses so they are also logged

martinellison commented:
I've added some code so it only journals some commands (anything that changes the game state), but not others (such as quit or print). This tidies the journal (and removes the blank lines). It was easier to implement this by adding and extra flag to the `command` struct, so the initialization now has the extra flag on each command entry.

I have also modified various code so that more command responses are written to the journal. Commands changed include `yesno` and `increase` for example. Hopefully, all changes to the game state will be journalled now.  I haven't journaled the explanation responses (`-> Rulebook`) but this can be done by changing `printf` calls to `pm` calls as required.

thexhr commented:
FYI: I am traveling the next 2 weeks, so expect delays or no merges/reviews at all.

thexhr commented on isscrolls.h:
```diff
@@ -155,7 +155,12 @@ int get_si(void);
-void write_journal_entry(char const * const what);
+extern FILE *journal_file;
+void print_to_journal(const char *, ...);
+void print_to_journal_v(const char *, va_list *);
```
This does not compile on OpenBSD with the following message:

```
cc -g3 -ggdb -pipe -fdiagnostics-color -Wno-unknown-warning-option -Wpedantic -Wall -Werror-implicit-function-declaration -Wno-format-truncation -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wshadow -Wpointer-arith -Wcast-qual -Wswitch-enum -Wuninitialized -Wformat-security -Wformat-overflow=2 -Wextra -I/usr/local/include `pkg-config --cflags json-c` -o rolls.o -c rolls.c
In file included from rolls.c:23:
./isscrolls.h:160:39: error: unknown type name 'va_list'; did you mean '__va_list'?
void print_to_journal_v(const char *, va_list *);
                                      ^~~~~~~
                                      __va_list
/usr/include/machine/_types.h:126:27: note: '__va_list' declared here
typedef __builtin_va_list       __va_list;
                                ^
1 error generated.
```

I could fix it with the proposed __va_list fix. However, this resulted in non-compiling code on Linux.

Simple but ugly fix would be to use `#idef` :/ Do you have a better way to address this?

martinellison replied:
I have now explicitly included `stdargs.h` in `isscrolls.h`. `stdargs.h` is supposed to declare `va_list` in C89 and later. Does that help?

martinellison replied:
@thexhr Have you had a chance to check whether this builds for you now?

thexhr replied:
Forget to check again after travel. Thanks for the reminder. Works on OpenBSD as well.

thexhr commented:
I merged your code into master, thanks again. Now let's calm development down a bit and give users some time to checkout latest master to play with the new features and find any potential bugs before I tag a new release.

Feel free to advertise your work on Reddit or other Ironsworm platforms you might use.

martinellison commented:
Are we ready to close the issue (#14) then?

I want to do some user testing for a while to see what the user interface
needs.

On Tue, 15 Apr 2025, 22:18 Matthias Schmidt, ***@***.***>
wrote:

> I merged your code into master, thanks again. Now let's calm development
> down a bit and give users some time to checkout latest master to play with
> the new features and find any potential bugs before I tag a new release.
>
> Feel free to advertise your work on Reddit or other Ironsworm platforms
> you might use.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/thexhr/isscrolls/pull/18#issuecomment-2805348251>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEA3AQF36FTLKEQIO3LR5D2ZUILVAVCNFSM6AAAAABZS5H2TSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBVGM2DQMRVGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> *thexhr* left a comment (thexhr/isscrolls#18)
> <https://github.com/thexhr/isscrolls/pull/18#issuecomment-2805348251>
>
> I merged your code into master, thanks again. Now let's calm development
> down a bit and give users some time to checkout latest master to play with
> the new features and find any potential bugs before I tag a new release.
>
> Feel free to advertise your work on Reddit or other Ironsworm platforms
> you might use.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/thexhr/isscrolls/pull/18#issuecomment-2805348251>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEA3AQF36FTLKEQIO3LR5D2ZUILVAVCNFSM6AAAAABZS5H2TSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBVGM2DQMRVGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
